### ISSUE TITLE
Make plot_scatter() available to user

### ISSUE THREAD

**Issue by nh3** [scverse/scanpy#617]
Currently, if users are to make scatter plots based on non-standard embeddings, they use `sc.pl.scatter()`, which is functionally similar to `plot_scatter()` but not exactly the same. Be able to plot non-standard embeddings alongside standard embeddings (e.g. using `sc.pl.umap()`) in exactly the same size and style is often desired. For this reason, do you think it's ok to export `plot_scatter()` or even export it as `sc.pl.scatter()`? Many thanks!

**Comment by ivirshup** [comment]
I agree that `plot_scatter` should be exported, and with a different name. Since `sc.pl.scatter` does different things, maybe it could be exported as  something like `sc.pl.dimred` or `sc.pl.obsm`?

**Comment by falexwolf** [comment]
The question is whether we want two scatter functions. Two will lead to simpler code. The current `pl.scatter` function is a relict from earlier times and could be simplified by dropping support for `basis`.

On the other hand, it might not be so much work to add support for `x` and `y` in `plot_scatter`, in which case both functions should do the same thing.

If we decide we want two functions, the second one should be `pl.scatter_embedding`, I'd say, as it's a method to visualize results of the embedding algorithms. I'm also ok with `pl.scatter_obsm` (`dimred` is something that is basically not used in the language of the docs, and I think it's a pretty misguided name of a class of functions anyway). The first one can remain `pl.scatter` and is similar to `plt.scatter` or `sns.scatter`.

**Comment by nh3** [comment]
Sorry to be unresponsive for a while. I think exporting `plot_scatter` as `pl.scatter_embedding` and keeping `pl.scatter` sounds a like a simple and good solution!

**Comment by flying-sheep** [comment]
@VolkerBergen has something similar in scvelo, we should definitely go for his design!

**Comment by VolkerBergen** [comment]
In scvelo I have decided to handle everything within one single `pl.scatter` module, where you can pass anything to `basis`, `x`, `y` and `color` from obs/var keys to arrays to lists of such. The implementation is rather simple and condensed, and @flying-sheep I'm happy to support you if (partly) merging into scanpy sounds reasonable to you.

You find some exemplary use cases in this notebook: https://scvelo-notebooks.readthedocs.io/Pancreas.html

**Comment by ivirshup** [comment]
@flying-sheep, didn't @fidelram implement this as `sc.pl.embedding` in #794? I think it got discussed in a different issue (https://github.com/theislab/scanpy/issues/762#issuecomment-517978906), so this didn't get closed.

**Comment by flying-sheep** [comment]
Yes, but I’m not happy about the spaghetti code in pl.scatter. We should make pl.embedding and pl.scatter share most of their code, and make that code as simple as @VolkerBergen’s.

**Comment by VolkerBergen** [comment]
Happy to discuss what can be integrated from scvelo's `pl.scatter` into scanpy or how scvelo's codebase can be used.. just to mention some of the features that may also be interesting for scanpy:
- (`x`, `y`) is `str` key of (var_names, var_names), (var, var), (obs, obs), (array, array), (obs, var_names), where I find particularly passing arrays to be very convenient.
- `basis` from obsm (what is the reason for having an additional `pl.embedding`?) or var_names (on layer1 vs layer2, e.g. spliced vs. unspliced).
- `color` is `str` key of obs, var, layers or directly pass an array (which I find very convenient)
while each of these can also be a list/tuple of `str` or arrays. 

Further, we 'beautified' the colorbar, ticks etc. and added some functionality such as plotting a lin.reg line or polynomial fit of any degree directly on top of the scatterplot, show histogram/density along x and y axes, added `dpi` and `figsize` attributes and **kwargs for all other matplotlib-specific attributes such as `vmin`/`vmax`. 
Apart from these it entails all functionality of scanpy's `pl.scatter`. It turned out to be very convenient to have pretty much everything within one single `pl.scatter` module, not matter whether you want to visualize an embedding, any user-specified arrays colored by clusters, or visualize a gene trend along a pseudotime.

I'd start of with the general question of whether incorporating some of these functionalities into scanpy's `pl.scatter` that may be useful, or whether re-implementing it based on scvelo's `pl.scatter` codebase makes more sense.

**Comment by fidelram** [comment]
I welcome @VolkerBergen ideas about plot scatter. I have used the scvelo version of scatter and works quite well and always thought that we could integrate this. Our comprehensive collection of tests related to embeddings should facilitate the recreation of the current functionality using a scatter module.

As @flying-sheep points out we have a mess with respect to `pl.scatter` and `pl.embeddings` and would be great to unify the code. Currently,  `pl.scatter` is used to plot two genes or any two variables like in `sc.pl.highly_variable_genes`. `pl.embedding` takes x,y (and z if 3D) from `.obsm` while adjusting color and size depending on given parameters.

When I started working on the plotting functions I didn't touch `pl.scatter` which remains quite convoluted and hard to follow.

**Comment by VolkerBergen** [comment]
@stefanpeidli is currently testing whether scVelo `pl.scatter` entails all scanpy functionality. @fidelram Can you give me a very brief outline of what functionality you added in `pl.embedding` so that I can account for these as well?

**Comment by VolkerBergen** [comment]
Tested scvelo's vs scanpy's `pl.scatter` (will share test modules + notebooks later). scvelo entails scanpy's functionality (and adds a couple extensions as explained above), except for the following:
- we don't have `left_margin` and `right_margin` attributes (don't think they're necessary).
- point `size` convention acc. to scanpy (if settings._rcParams_style == 'scanpy') else slightly adjusted.
- when `basis` is in `.var_names`, then an unspliced/spliced phase portrait is plotted (I guess that's not needed in scanpy?)

Further we have some more defaults:
- if `basis` is None, then use a default basis in the given order if available: umap, tsne, pca
- if `color` is None, then use a default color in the given order if available: clusters, louvain
- if `frameon` is None, then only set frame if it is not embedding and axes values do matter.
- if `color_map` is None, then use 'viridis_r' if vals in [0,1], else matplotlib's default.

Further, these can be used interchangeably: `size` and `s`, `color` and `c`, `color_map` and `cmap`.

**Comment by fidelram** [comment]
@VolkerBergen on the top of my head here are some features that were added recently:
 * adjustment of multiple plots distance using `wspace` and `hspace` (indeed I have never used `left_margin` and `right_margin` which I think are not necessary)
 * multiple plots by a combination of `color` and `components`. E.g. This will produce four plots: `sc.pl.pca(adata, color=['CD14', 'CD8'], components=['1,2', '2,3']` 
 * support for layers
 * when the marker size given by the user is a vector, this is also subjected to any sorting of the cells (eg. `sort_order=True`, `groups=['<group name']` )
 * added the option to outline the plot using the parameter `add_outline` #794 
* added vmin and vmax as percentiles #794 
 * when using the `groups` parameter, plot those groups on top #891

**Comment by VolkerBergen** [comment]
Thanks @fidelram. That is all included within scvelo's `scatter`. However, `vmin` and `vmax` are not given as explicit arguments, but are passed via `**kwargs` to matplotlib directly. Will add few more details to the docs and then open a PR soon.

**Comment by ivirshup** [comment]
This has addressed by the addition of the `embedding` function.
